Add app name to challenge progress tracker

convolu opened the pull request:
Also make it absolutely clear that this class is a singleton

Related to https://github.com/KanoComputing/peldins/issues/2248

convolu commented:
@tombettany @alex5imon @pazdera Could you please review?

pazdera commented:
Looks good! One thing that comes in mind: should we raise when the app name hasn't been set at the time when the first tracking data arrive? If the event's don't make sense without it ...

convolu commented:
Usually you should be able to infer the app from other entries (such as  `challenge_group`). Having said that I think it is a good idea, I will add it to make sure we catch misuse as early as possible

convolu commented:
@pazdera I've updated as per your suggestion. Could you take another look please?

Ealdwulf commented:
It's a bit strange that we have an 'app_session_id' but this doesn't seem to correspond to the session entry that eventually turns up in the tracking events. I would have expected that to be the way the challenge event gets associated with the app.

pazdera commented:
@Ealdwulf That's a good point, I haven't thought of that. We should probably keep those in sync, what we use for session names and these app_ids.

We seem to be using `get_program_name()` from utils for app names in the tracker: https://github.com/KanoComputing/kano-profile/blob/master/kano_profile/tracker.py#L320

Maybe we could keep the sessions and progress reports closer via some sort of id (as Alex suggests). Now Vincent will have to use the timestamps to make the connection between progress made and sessions started.

tombettany commented on kano_profile/challenge_progress_tracker.py:
```diff
@@ -19,13 +19,45 @@ class ChallengeProgressTracker(object):
+        else:
+            raise ValueError('App name must not be empty')
+
+    def _check_name_is_set(self):
```
Given that we run this at the start of every function, it might be nice to make this a decorator. It is fine as it is if it is too much effort.

tombettany commented:
:+1:

convolu commented:
@pazdera @Ealdwulf I have updated the PR as per your suggestions. The app session gets a session id that then is acquired by the challenge tracker and added to the tracking events. Could you please take another look?

pazdera commented on kano_profile/challenge_progress_tracker.py:
```diff
@@ -19,12 +22,46 @@ class ChallengeProgressTracker(object):
         self._chal_no = None
         self._chal_outcome = None
-        self._app_session_id = str(uuid4())
+        self._app_session_id = get_session_unique_id(app_name, getpid())
```
Would it be appropriate to log a warning if this doesn't return anything (i.e., the session tracker wasn't initialised yet). Or do you see use cases that won't be tied to a session?

convolu replied:
Hm, I think that we could add the warning anyway

Ealdwulf commented:
Looks good to me @convolu :+1:

convolu commented:
@pazdera I have added a warning when the session id is not found.

pazdera commented:
Thanks @convolu. Otherwise, :shipit:!
